### ISSUE TITLE
Fargate agent with `containerDefinitions_environment` throws `string indices must be integers` error

### ISSUE THREAD

**Issue by oleksandr** [PrefectHQ/prefect#2863]
## Description

Fargate agent is executed as is and configured through env variables only:

```yaml
    environment:
      - PREFECT__BACKEND=server
      - PREFECT__SERVER__HOST=http://x.x.x.x
      - PREFECT__LOGGING__LEVEL=DEBUG
      - PREFECT__FLOWS__CHECKPOINTING=true
      - AWS_ACCESS_KEY_ID=...
      - AWS_SECRET_ACCESS_KEY=...
      - REGION_NAME=eu-central-1
      - containerDefinitions_logConfiguration=.{..
      - containerDefinitions_repositoryCredentials={...
      - containerDefinitions_environment={"AWS_S3_BUCKET":"DEMO123"}}
```

Upon the start looks like config is picked up by agent:

```
...
prefect-fargate-agent_1  | [2020-06-25 08:38:13,441] DEBUG - dev-layout-fargate-agent | Container definition: environment from environment variable
prefect-fargate-agent_1  | [2020-06-25 08:38:13,441] DEBUG - dev-layout-fargate-agent | Container definition: logConfiguration from environment variable
prefect-fargate-agent_1  | [2020-06-25 08:38:13,442] DEBUG - dev-layout-fargate-agent | Container definition: repositoryCredentials from environment variable
prefect-fargate-agent_1  | [2020-06-25 08:38:13,488] DEBUG - dev-layout-fargate-agent | Launch type: FARGATE
...
```

But right before the task is created in AWS the logs show:

```
prefect-fargate-agent_1  | [2020-06-25 08:39:15,483] DEBUG - dev-layout-fargate-agent | Using image registry.gitlab.com/********:master for task definition
prefect-fargate-agent_1  | [2020-06-25 08:39:15,483] DEBUG - dev-layout-fargate-agent | Removing static Prefect envars from container_definitions_kwargs if exists
prefect-fargate-agent_1  | [2020-06-25 08:39:15,484] ERROR - dev-layout-fargate-agent | Logging platform error for flow run aa680949-2999-4c54-b06a-d32bb112e977
prefect-fargate-agent_1  | [2020-06-25 08:39:15,532] ERROR - dev-layout-fargate-agent | Error while deploying flow: TypeError('string indices must be integers')
prefect-fargate-agent_1  | [2020-06-25 08:39:15,532] DEBUG - dev-layout-fargate-agent | Completed flow run submission (id: aa680949-2999-4c54-b06a-d32bb112e977)
```

And the task is actually NOT submitted.

## Expected Behavior

The `AWS_S3_BUCKET` to be available in the taskDefintion's environment and passed properly to a running container.

## Environment

Running latest Prefect code from `master`.

**Comment by joshmeek** [comment]
@oleksandr I believe `environment` expects a list of dictionaries (seen [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ecs.html#ECS.Client.register_task_definition)). Change that to a list and see if it fixes the issue 🙂

**Comment by oleksandr** [comment]
@joshmeek bummer. indeed. all works fine! can close Fargate chapter :-) Thanks a lot!

**Comment by Giaco9** [comment]
Hi,
 I think we have the same problem but I don't understand where the error is. I deploy the Fargate agent using cloud formation on ECS and this is the container definition I use. In the environment section I declared `containerDefinitions_environment` as declared in the register_task_defintion boto3 API.

```
ContainerDefinitions:
    - Name: 'agent'
      Essential: true
      Command: ['prefect', 'agent', 'start', 'fargate', 'enable_task_revisions=true']
      Image: !Sub
        - 'prefecthq/prefect:${version}'
        - version: !If [ cIsPrefectServerVersionLatest, 'all_extras', !Sub 'all_extras-${pPrefectServerVersion}' ]
      MemoryReservation: 128
      Environment:
		- Name: PREFECT__CLOUD__API
		  Value: 'https://apollo-prefect-server.srv-stage.cloudacademy.xyz/graphql'
		- Name: PREFECT__BACKEND
		  Value: 'server'
		- Name: REGION_NAME
		  Value: !Ref AWS::Region
		- Name: executionRoleArn
		  Value: !GetAtt 'rAgentExecutionRole.Arn'
		- Name: taskRoleArn
		  Value: !GetAtt 'rAgentTaskRole.Arn'
		- Name: cluster
		  Value: !Ref pEcsCluster
		- Name: networkConfiguration
		  Value: !Sub
		    - '{"awsvpcConfiguration": {"assignPublicIp": "ENABLED", "subnets": ["${subnets}"], "securityGroups": ["${securityGroups}"]}}'
		    - subnets: !Join ['","', !Ref pFargateSubnetIds]
		      securityGroups: !Join ['","', !Ref pFargateSecurityGroupIds]
		- Name: cpu
		  Value: 256
		- Name: memory
		  Value: 512
		- Name: enableECSManagedTags
		  Value: true
		- Name: containerDefinitions_environment
		  Value: !Sub
		    - '
		        [{
		          "name": "DB_JSON_CONNECTION_SECRET_ARN",
		          "value": "${DBJsonConnection}"
		        }, {
		          "name": "Environment",
		          "value": "${Environment}"
		        }]
		      '
		    - DBJsonConnection: !Ref rPrefectAgentContainerDefinitionSecretsSecret
		      Environment: !Ref pEnvironment
      LogConfiguration:
        LogDriver: 'awslogs'
        Options:
          awslogs-group: !Ref 'rAppLogsGroup'
          awslogs-region: !Ref 'AWS::Region'
          awslogs-stream-prefix: 'prefect-agent'
```

After I added the `containerDefinitions_environment` environment variable the agent throws the error `Error while deploying flow: TypeError('string indices must be integers')`.

Do you have any idea of why this is happening?

Thank you!

**Comment by AlessandroLollo** [comment]
After some tries and errors we finally got it working with the following setup
`
- Name: containerDefinitions_environment
              Value: !Sub
                - '[{"name": "DB_JSON_CONNECTION_SECRET_ARN", "value": "${DBJsonConnection}"}, {"name": "Environment", "value": "${Environment}"}]'
                - DBJsonConnection: !Ref rPrefectAgentContainerDefinitionSecretsSecret
                  Environment: !Ref pEnvironment
`

Basically, the value of `containerDefinitions_environment` should not contain line feed or carriage return chars.
Is this the intended behaviour or is it a bug (Prefect or boto3)?
